File: AWSIOTMQTT/app/src/main/java/com/dilip/awsiotmqtt/MainActivity.java
package com.dilip.awsiotmqtt;

import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.amazonaws.services.iot.client.AWSIotException;
import com.amazonaws.services.iot.client.AWSIotMqttClient;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        String clientEndpoint = "<prefix>-ats.iot.<region>.amazonaws.com";   // use value returned by describe-endpoint --endpoint-type "iot:Data-ATS"
        String clientId = "SoilMoistureSensor";                              // replace with your own client ID. Use unique client IDs for concurrent connections.
        String certificateFile = "<certificate file>";                       // X.509 based certificate file
        String privateKeyFile = "<private key file>";                        // PKCS#1 or PKCS#8 PEM encoded private key file

// SampleUtil.java and its dependency PrivateKeyReader.java can be copied from the sample source code.
// Alternatively, you could load key store directly from a file - see the example included in this README.
        SampleUtil.KeyStorePasswordPair pair = SampleUtil.getKeyStorePasswordPair(certificateFile, privateKeyFile);
        AWSIotMqttClient client = new AWSIotMqttClient(clientEndpoint, clientId, pair.keyStore, pair.keyPassword);

// optional parameters can be set before connect()
        try {
            client.connect();
        } catch (AWSIotException e) {
            throw new RuntimeException(e);
        }
        Toast.makeText(this, "Connected!", Toast.LENGTH_SHORT).show();
        Log.d("MQTT", "\"Connected!");
    }
}

File: AWSIOTMQTTConnection/app/src/main/java/com/dilip/awsiotmqttconnection/MainActivity.kt
package com.dilip.awsiotmqttconnection

import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import com.amazonaws.mobileconnectors.iot.AWSIotMqttClientStatusCallback
import com.amazonaws.mobileconnectors.iot.AWSIotMqttQos

class MainActivity : ComponentActivity() {

    private lateinit var awsIotClient: AWSIoTClient

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // ... (your Jetpack Compose UI content here)

        val clientEndpoint = "<prefix>-ats.iot.us-west-2.amazonaws.com"
        val clientId = "sdk-java-23"
        val certificateFile = "athing.cert.pem"
        val privateKeyFile = "athing.private.key"

        // Optional: Implement custom logic for credentials provider
        val credentialsProvider = // Your credentials provider

        // Load keystore and password
        val keyStorePasswordPair = KeyStoreUtil.getKeyStorePasswordPair(certificateFile, privateKeyFile)

        awsIotClient = AWSIoTClient(clientEndpoint, clientId, keyStorePasswordPair, credentialsProvider)

        // Connect to AWS IoT
        val connectionStatus = awsIotClient.connect()
        if (connectionStatus == AWSIotMqttClientStatusCallback.AWSIotMqttClientStatus.Connected) {
            Log.d("TAG", "Connected to AWS IoT")
            val topic = "sdk/test/java"
            val payload = ""// Your payload data (e.g., JSON string)
                awsIotClient.publishString(topic, AWSIotMqttQos.QOS0, payload)
            Log.d("TAG", "Message sent")
            // ... (other functionality in your activity)
        } else {
            Log.e("TAG", "Connection error")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        awsIotClient.disconnect()
    }
}
